doc: Add more useful information to Glossary.md

HarshithaKP opened the pull request:
<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [ ] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [ ] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

<!--
Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

joyeecheung commented on glossary.md:
```diff
@@ -1,16 +1,82 @@
-  and tampering with built-ins.
+* bootstrap: Early phase in the Node.js process startup - that sets up
+  the Node.js execution environment and loads the internal modules.
+* code cache: A chunk of memory where the compiled JS code is stored.
```
```suggestion
* code cache: A chunk of bytes storing compiled JS code and its metadata
```

It does not necessarily reside in memory (can be written to disk)

HarshithaKP replied:
@joyeecheung, thanks. Addressed the change, PTAL.

joyeecheung commented on glossary.md:
```diff
@@ -1,16 +1,82 @@
+  callable from JS.
+* primordials: Pristine built-ins that are not effected by prototype
+  pollution and tampering with built-ins.
+* snapshot: Native object code compiled from the JS APIs of Node.js
```
```suggestion
* snapshot: when referring to the V8 startup snapshot, a chunk of bytes
  containing serialized V8 heap, which can be deserialized to restore
  the state of the V8 engine.
```

There is also the heap snapshot BTW.

HarshithaKP replied:
@joyeecheung, ditto.

joyeecheung commented on glossary.md:
```diff
@@ -1,16 +1,84 @@
-* LGTM: "Looks good to me", commonly used to approve a code review.
-* PTAL: Please take a look.
-* RSLGTM: "Rubber-stamp looks good to me". The reviewer approving without doing
+* ABI: Application Binary Interface - contracts for internal
```
Do we really need entries that can be easily looked up in Wikipedia or OS textbooks and widely used in the software world in general? The same goes to things like CLI, EOF, ELO, RAII, etc.

HarshithaKP replied:
@joyeecheung, (a) the problem this gloassry addresses is difficulty in discovering meaning of abbreveations, and solution is to keep those in one place for easy reference; based on our usage of those, not based on how trivial those are? (b) I did a grep in the code base and found repeated usage of these, (c) these items were proposed in the original issue by maintainers

starkwang commented:
I'm +1 on adding more information to Glossary.md but -1 on adding general words like `OSX`, `ECMA ` or `MDN`.

HarshithaKP commented:
@starkwang, `general` is a relative term right? for folks who don't know the meaning of those will also don't know where to search for it?

my general request is to look at the list from a new comer's angle, not from an experienced node.js programmer's.

joyeecheung commented:
> for folks who don't know the meaning of those will also don't know where to search for it?

I'd suggest we give links to "where to search for it" instead of trying to explaining them ourselves.  For example, the explanation to ABI here:

> ABI: Application Binary Interface - contracts for internal API invocations.

seems misleading to me - API is a concept at the level of source code, which is relevant in *writing programs*, not in *loading binaries*. On the other hand [Wikipedia](https://en.wikipedia.org/wiki/Application_binary_interface) does a better job explaining it:

> An ABI defines how data structures or computational routines are accessed in machine code, which is a low-level, hardware-dependent format; in contrast, an API defines this access in source code, which is a relatively high-level, hardware-independent, often human-readable format. A common aspect of an ABI is the calling convention, which determines how data is provided as input to or read as output from computational routines; examples are the x86 calling conventions.

> my general request is to look at the list from a new comer's angle, not from an experienced node.js programmer's

While I agree that the list exists to serve new comers, I don't think we need to extend the list to include terms that are not particularly different in the context of Node.js and that most people learn from elsewhere. For instance `code cache` and `snapshot` carry special meanings in this repository, so it makes sense to include them here, while `ABI` or `RAII` doesn't - those are concepts regularly taught in courses/books/tutorials/videos/whatever about systems and/or C/C++, and it's better that people learn them from those sources instead of here. Otherwise we end up recreating an index of terms of a textbook about system or C/C++ here, which is a good effort but IMO not an effort that belongs in this repo.

gireeshpunathil commented:
one advantage of this approach (as opposed to linking to external sites) is, meaning can be interpreted in our usage context. example: https://github.com/nodejs/node/blob/0c3c0e7184c4ca625148c14d61d8e9438a1b69f2/src/node_version.h#L78-L80

here, we are not really using this term to refer to hardware architectures or subroutine linkage conventions; but to a protocol by which  node interacts with the the dependent modules? so instead of taking to the general literature of its definition, history etc. a short description that will satisfy the need for some one reading through the code will be more effective?

another example: https://github.com/nodejs/node/blob/0c3c0e7184c4ca625148c14d61d8e9438a1b69f2/src/node_options.cc#L541-L544

here, the reader may not want to know the complete story of https://en.wikipedia.org/wiki/Read%E2%80%93eval%E2%80%93print_loop - as at that point they may not be interested in that; but just to fill the gap of their understanding (a simple expansion) and move on with the option parsing?

we do have precedence here in https://github.com/nodejs/node/blob/0c3c0e7184c4ca625148c14d61d8e9438a1b69f2/CPP_STYLE_GUIDE.md as well as https://github.com/nodejs/node/blob/0c3c0e7184c4ca625148c14d61d8e9438a1b69f2/src/README.md that provide `customized knowledge` on C++ usage in the code base.

and technically too I find this approach correct - sticking to the definition of glossary: a `brief` dictionary of `used` words that are uncommon.

joyeecheung commented:
@gireeshpunathil I agree that providing customized knowledge about these terms in the context of Node.js would be useful (for example, by `REPL` we mean "the thing you see when you run `node` in the terminal") , however by putting things into context it would go against the goal of keeping the dictionary brief. For example, IMO if you do not know what `ABI` or `RAII` is, expanding them to `Application Binary Interface` or `Resource Aquisition Is Intialization` is still not sufficient for you to understand the comments mentioning them and keep reading/editing the code. To put those terms into the context of Node.js we need to explain by giving examples. Also, that's not what this PR is doing right now - it is giving general explanations about these terms.

gengjiawen commented on glossary.md:
```diff
@@ -1,16 +1,82 @@
+* V8: JavaScript engine that is embedded in Node.js.
+* VM: Virtual Machine, in the context of abstracting instructions,
+  not to be confused with H/W level or O/S level Virtual machines.
+* WASI: Web Assemby System Interface.
```
Link to this https://github.com/WebAssembly/WASI will be more informative.

gengjiawen commented on glossary.md:
```diff
@@ -1,16 +1,82 @@
+* ASYNC: Asynchronous
+* BE: Big Endian - Byte order style in a multibyte data.
+  (Opposite of LE, Little Endian)
+* CI: Continuous Intigration.
```
typo `Intigration`

HarshithaKP replied:
@gengjiawen, thanks. Fixed it.

HarshithaKP commented:
@joyeecheung and @gengjiawen , addressed your concerns by removing explanation and providing external links wherever it is applicable. Please have a look and let me know if you are good with that.

HarshithaKP commented:
@gengjiawen -

> gengjiawen self-requested a review 2 days ago

did you mean to keep this for your review-todo list,

or

did you mean to dismiss your earlier change-request?

gengjiawen commented:
> @gengjiawen -
> 
> > gengjiawen self-requested a review 2 days ago
> 
> did you mean to keep this for your review-todo list,
> 
> or
> 
> did you mean to dismiss your earlier change-request?

I mean to dismiss your earlier change-request.

Also I am -1 on adding `MDN` `MAC` and `RAII` such C++ staff.

HarshithaKP commented:
@gengjiawen, can you please explain your reason? Is it that you think those do not need explanation, or thier explanations are not accurate?

joyeecheung commented on glossary.md:
```diff
@@ -1,16 +1,82 @@
-* RSLGTM: "Rubber-stamp looks good to me". The reviewer approving without doing
+* ABI: [Application Binary Interface](https://en.wikipedia.org/wiki/Application_binary_interface)
+* ASAP: As Soon As Possible.
+* ASM: Assembler - General term for language translator, especially
```
To clarify, I still think this criteria of inclusion in this list seems puzzling - if we are explaining what assembler is, why don't we also explain compiler, linker...and to that end, computer?

gireeshpunathil replied:
@joyeecheung -

> why don't we also explain about compiler, linker...and to that end, computer?

Is this a joke, or sarcasm?

criteria of inclusion in this list is very simple IMO (I proposed majority of these items) - presence of their usage in our code base and that are not obvious to a user who are normally skilled in the art.

first criteria is objective, measurable:

https://github.com/nodejs/node/blob/c6ed01cf6d0d54f8de862fd41413e5cbac2a8ba4/configure.py#L744

second criteria is subjective, and my assertions are based on my interactions with users, mostly through the help repo: Most users there know what a computer is, but may not know what `ASM` is.

joyeecheung replied:
>  warn('''No acceptable ASM compiler found! 

To me in this particular example it looks like the text should be changed to `assembler` instead, or something more informative like `.asm` compiler (I am not even sure if that's a proper term), instead of adding a glossary for it. `ASM compiler` AFAIK, is not widely used in this repo when people talk about things either.

HarshithaKP replied:
@joyeecheung, removed `ASM` from the document.

joyeecheung commented on glossary.md:
```diff
@@ -1,16 +1,80 @@
+* IPC: Inter Process Communication.
+* JIT: Just In Time - General term for dynamic compiler in
+  managed runtimes.
+* JS/C++ boundary: A layer that bridges between JS APIs and the C++
```
I don't think is accurate - when we talk about `JS/C++ boundary` we are usually referring to the boundary between V8's runtime and the JS code execution which are commonly crossed when invoking JS functions with C++ linkage (e.g. interceptors), i.e. something like this

```
    frame #11: 0x00000001009703db node`Builtins_InterpreterEntryTrampoline + 187
    frame #12: 0x00000001009703db node`Builtins_InterpreterEntryTrampoline + 187
    frame #13: 0x00000001009703db node`Builtins_InterpreterEntryTrampoline + 187
    frame #14: 0x000000010096dfda node`Builtins_JSEntryTrampoline + 90
    frame #15: 0x000000010096ddb8 node`Builtins_JSEntry + 120
```

Instead of the conceptual layer in the implementation that abstracts platform capabilities. The boundary could very well exist when we are reaching into the C++ to do something that has nothing to do with the platform, e.g. URL parsing.

HarshithaKP replied:
Thanks, Changed it. PTAL.

joyeecheung commented on glossary.md:
```diff
@@ -1,16 +1,80 @@
+* LTS: Long Term Support.
+* MDN: Mozila Development Network - A vast and authentic
+  documentation on JavaScript.
+* OOB: Out Of Bounds - Used in the context of TCP data transport.
```
I don't think this is where OOB is mostly used - we primarily talk about OOB in array(memory) access

HarshithaKP replied:
Changed it according to suggestion. PTAL

joyeecheung commented on glossary.md:
```diff
@@ -1,16 +1,80 @@
+  documentation on JavaScript.
+* OOB: Out Of Bounds - Used in the context of TCP data transport.
+* OOM: Out Of Memory.
+* OSX: Mac OS.
```
This entry seems somewhat odd - why assume more people know about macOS than OS X? And it doesn't seem right to me for us to explain about Apple's branding strategy here

HarshithaKP replied:
Removed it.

joyeecheung commented on glossary.md:
```diff
@@ -1,16 +1,80 @@
+* PPC: Power PC.
+* PTAL: Please Take A Look.
+* RAII: [Resource Aquisition Is Intialization](https://en.wikipedia.org/wiki/Resource_acquisition_is_initialization)
+* REPL: Read Evaluate Print Loop.
```
This does not really explain what it is in the context of Node.js - I think people who don't know what REPL might still be puzzled after reading this.

HarshithaKP replied:
Explained it. PTAL.

joyeecheung commented on glossary.md:
```diff
@@ -1,16 +1,80 @@
+* RFC: Request For Comments.
+* RSLGTM: Rubber-Stamp Looks Good To Me. The reviewer approving without doing
   a full code review.
+* RSS: Resident Set Size.
```
It does not seem to be meaningful to just expand the word. A link to https://en.wikipedia.org/wiki/Resident_set_size would be more useful

HarshithaKP replied:
Thanks, Linked it.

joyeecheung commented on glossary.md:
```diff
@@ -1,16 +1,80 @@
+* RSLGTM: Rubber-Stamp Looks Good To Me. The reviewer approving without doing
   a full code review.
+* RSS: Resident Set Size.
+* SMP: Symmetric Multi Processor.
```
Where is this used? I don't recall seeing anyone talking about this in Node.js.

HarshithaKP replied:
It's not used. Removed it.

joyeecheung commented on glossary.md:
```diff
@@ -1,16 +1,80 @@
   [TSC](./GOVERNANCE.md#technical-steering-committee).
-* WIP: "Work In Progress" - e.g. a patch that's not finished, but may be worth
-  an early look.
+* V8: JavaScript engine that is embedded in Node.js.
```
It seems somewhat funny to define `V8` as `JavaScript engine that is embedded in Node.js` which may sends you the impression (provided that you don't know what it is) that this is a code name for a component developed for Node.js

HarshithaKP replied:
Changed the definition, PTAL.

joyeecheung commented on glossary.md:
```diff
@@ -1,16 +1,80 @@
-* WIP: "Work In Progress" - e.g. a patch that's not finished, but may be worth
-  an early look.
+* V8: JavaScript engine that is embedded in Node.js.
+* VM: Virtual Machine, in the context of abstracting instructions,
```
I think in this code base the most commonly used meaning of `vm` is actually the `vm` builtin module, which is not really a Virtual Machine but just an unfortunate name for something it is not..and describing VM this way does not seem to help clarifying that misconceptioin.

HarshithaKP replied:
Thanks, changed it accordingly and linked it. PTAL.

joyeecheung commented on glossary.md:
```diff
@@ -1,16 +1,80 @@
+  helpers that implements / abstracts platform capabilities.
+* LGTM: Looks good to me - commonly used to approve a code review.
+* LTS: Long Term Support.
+* MDN: Mozila Development Network - A vast and authentic
```
Why not link to it directly?

HarshithaKP replied:
Thanks, linked it.

joyeecheung commented on glossary.md:
```diff
@@ -1,16 +1,80 @@
+* ASYNC: Asynchronous
+* BE: Big Endian - Byte order style in a multibyte data.
+  (Opposite of LE, Little Endian)
+* CI: Continuous Integration.
```
This does not seen to be useful for people who don't know what CI is

HarshithaKP replied:
Linked it to Wikipedia explanation.

joyeecheung commented on glossary.md:
```diff
@@ -1,16 +1,80 @@
-* RSLGTM: "Rubber-stamp looks good to me". The reviewer approving without doing
+* ABI: [Application Binary Interface](https://en.wikipedia.org/wiki/Application_binary_interface)
+* ASAP: As Soon As Possible.
+* ASYNC: Asynchronous
```
Why is `SYNC` left out if `ASYNC` needs explanation?

HarshithaKP replied:
Removed it.

joyeecheung commented on glossary.md:
```diff
@@ -1,16 +1,80 @@
+* CITGM: Canary In The Gold Mine - a smoke test unit in the repo that
+  contain popular npm modules.
+* CJS: Common JS.
+* CLDR: Common Locale Data Repository.
```
This does not seem to be useful for people who don't know what CLDR is

HarshithaKP replied:
Thanks, linked it to Wikipedia explanation.

HarshithaKP commented on glossary.md:
```diff
@@ -1,16 +1,80 @@
+* CJS: Common JS.
+* CLDR: Common Locale Data Repository.
+* CLI: [Command Line Interface](https://en.wikipedia.org/wiki/Command-line_interface)
+* CMD: Command.
```
Removed it.

HarshithaKP commented on glossary.md:
```diff
@@ -1,16 +1,80 @@
+  maintains reported security exposures.
+* ECMA: European Computer Manufacturers Association - A body that
+  governs JS spec among other things.
+* ENV: Environment - General term for execution environment.
```
Removed it.

HarshithaKP commented on glossary.md:
```diff
@@ -1,16 +1,80 @@
+  (when used within project documents).
+* ESM: ECMA Script Module.
+* ETW: Event Tracing for Windows.
+* FD: File Descriptor.
```
Removed it.

HarshithaKP commented on glossary.md:
```diff
@@ -1,16 +1,80 @@
+* FS: File System.
+* ICU: International Components for Unicode.
+* IPC: Inter Process Communication.
+* JIT: Just In Time - General term for dynamic compiler in
```
Thanks, changed it.

HarshithaKP commented:
@joyeecheung, Can you please take a look at my changes. I tried addressing most of your review comments.

jasnell commented:
Ping @joyeecheung

gireeshpunathil commented:
one more ping to @joyeecheung 
@HarshithaKP - can you pls rebase to resolve the file conflict?

HarshithaKP commented:
@gireeshpunathil sure

RedYetiDev commented:
Hi! It's been a few years since any activity on this PR.

If you're still interested in pursuing this, I suggest verifying that everything is still in working order after the long delay. Otherwise, you can always close the PR.

Nonetheless, your contribution is greatly appreciated!

> This is an attempt to resolve older PRs and issues

@nodejs/documentation

gireeshpunathil commented:
IMO, the info will be really useful especially for new comers, but the current form needs updates and validations.

RedYetiDev commented:
> IMO, the info will be really useful especially for new comers, but the current form needs updates and validations.

Agreed, but it might need a separate PR, as there have been a significant number of changes to the project since this PR.

@HarshithaKP do you still want to pursue this? A member of the team can also take over if you’d like.

gireeshpunathil commented:
I reviewed the change set now, and looks all valid to me. So if there are no objections and file rebased, it is ready to merge.

one lingering comment is why some of this can be found through general knowledge base such as google. Problem with wide search is lack of personalisation. For example when I searched "EOL" I got "EXTRA ORDINARY LEAVE". Here we are explaining the terms in context and that is largely beneficial to new comers.

@joyeecheung - do you have any observations?

RedYetiDev commented:
Great! I'm glad everything is still not conflicting!

RedYetiDev commented on glossary.md:
```diff
@@ -1,16 +1,79 @@
+* CJS: Common JS.
+* CLDR: [Common Locale Data Repository](https://en.wikipedia.org/wiki/Common_Locale_Data_Repository).
+* CLI: [Command Line Interface](https://en.wikipedia.org/wiki/Command-line_interface).
+* CVE: Common Vulnerebilities and Exposures - A database that
```
(Typo) CVE: Common **Vulnerabilities** and Exposures

RedYetiDev commented on glossary.md:
```diff
@@ -1,16 +1,79 @@
+* ETW: Event Tracing for Windows.
+* FFDC: First Failure Data Capture - Common terms for logs, traces
+  and dumps that are produced by default on program error.
+* FIPS:  Federal Information Processing Standards.
```
Double spacing is not needed

RedYetiDev reviewed (commented):
Few little nitpicks, but otherwise LGTM

joyeecheung commented on glossary.md:
```diff
@@ -1,16 +1,79 @@
+* CI: [Continuous Integration](https://en.wikipedia.org/wiki/Continuous_integration).
+* CITGM: Canary In The Gold Mine - a smoke test unit in the repo that
+  contain popular npm modules.
+* CJS: Common JS.
```
```suggestion
* CJS: CommonJS. In most cases, [CommonJS Modules](https://nodejs.org/api/modules.html#modules-commonjs-modules).
```

joyeecheung commented on glossary.md:
```diff
@@ -1,16 +1,79 @@
+* EOF: [End Of File](https://en.wikipedia.org/wiki/End-of-file)
+* EOL: End Of Line (when used within a program), End of Life
+  (when used within project documents).
+* ESM: ECMA Script Module.
```
```suggestion
* ESM: [ECMAScript Module](https://nodejs.org/api/esm.html#modules-ecmascript-modules)
```

joyeecheung commented on glossary.md:
```diff
@@ -1,16 +1,79 @@
+  of a program `at run time` rather than before execution.
+* JS/C++ boundary: It refers  to the boundary between V8's runtime and the
+  JS code execution which are commonly crossed when invoking JS functions with
+  C++ linkage (e.g. interceptors). The boundary could very well exist when we
```
```suggestion
  C++ linkage (e.g. `v8::FunctionCallback`). The boundary exist when we
```
interceptors are only very rarely used. `v8::FunctionCallback`s are a lot more common.

joyeecheung commented on glossary.md:
```diff
@@ -1,16 +1,79 @@
+* CVE: Common Vulnerebilities and Exposures - A database that
+  maintains reported security exposures.
+* ECMA: European Computer Manufacturers Association - A body that
+  governs JS spec among other things.
```
```suggestion
  governs the specification of the JavaScript language, among other things.
```

joyeecheung reviewed (approved):
Sorry for missing the pings, probably got lost during the hectic zero-COVID days..

RedYetiDev commented:
Moved to #52798. The original author is the Co-Author.
